### ISSUE TITLE
displayDialogAsync does not return 12007 when a dialog is already open on Mac

### ISSUE THREAD

**Issue by schwarzwald** [OfficeDev/office-js#3682]
According to the OfficeJs documentation the method Office.context.ui.displayDialogAsync should execute the callback function with an error code 12007 when a dialog is already open. 
On Mac the callback is executed as a success and a dialog instance is passed as a value.
This causes a problem when we're trying to open a second dialog immediately after closing the first one, because the second dialog instance is processing events from the first dialog which is being closed.

## Your Environment

* Platform: Mac
* Host: Outlook for Mac 16.76.2
 
## Expected behavior

The Office.context.ui.displayDialogAsync shoud execute the callback with a failure and error code 12007 when the dialog is still being closed. And the newly opened dialog should listen only to events which belong to that particular dialog.

## Current behavior

The Office.context.ui.displayDialogAsync executes the callback with a success and passes new dialog instance as a value before the previous dialog is fully closed.
Subsequently the new dialog receives close event from the old dialog.

## Steps to reproduce
```
let dialog: Office.Dialog; // let's assume we have an active dialog instance
dialog.close(); // call the close method which should asynchronously close the dialog

// try to open a new dialog immediately
Office.context.ui.displayDialogAsync(
  options,
  result => {
    if (result.status == Office.AsyncResultStatus.Failed) {
      if (result.error.code === 12007) {
        // dialog still open, try again...
        // ...
      }
    } else {
      let newDialog: Office.Dialog = result.value;
      // add a close handler to the new dialog
      newDialog.addEventHandler(Office.EventType.DialogEventReceived, () => {
        // this should be executed when the newDialog is closed
        // but it's executed after when the old dialog is finally closed
      });
    }
  }
})

**Comment by exextoc** [comment]
We tried the code at our end, and it is working as expected. The new dialog listens only to events which are meant for it and not for the first dialog. Can you please try the code snippet by using two different URLS's for both the dialogs, it will clarify better. Here is the sample code snippet:

var dialog;
var newDialog;
Office.context.ui.displayDialogAsync(URL1,
   function (asyncResult) {
       if (asyncResult.status === Office.AsyncResultStatus.Failed) {
            console.log(asyncResult.error.code + " = : " + asyncResult.error.message);
       } else {
            dialog = asyncResult.value;
            console.log("Dialog created");
            dialog.addEventHandler(Office.EventType.DialogEventReceived, () => {
        console.log("old dialog closed");
      });
            dialog.close();
            console.log("Dialog close called");
Office.context.ui.displayDialogAsync(URL2,
   function (asyncResult) {
       if (asyncResult.status === Office.AsyncResultStatus.Failed) {
            console.log("New dialog "+ asyncResult.error.code +"= : " + asyncResult.error.message);
       } else {
            newDialog = asyncResult.value;
        console.log("new dialog created");
            newDialog.addEventHandler(Office.EventType.DialogEventReceived, () => {
        console.log("new dialog closed");
      });

      }
});
      }
});

**Comment by schwarzwald** [comment]
Thanks for the reply. 
After further testing it seems the problem occurs only if the URLs of those two dialogs are the same except for the hash fragment.
In our case we have a react HashRouter for different dialogs and all URLs are in form (example) https://localhost:3000/index.html#dialog1   and https://localhost:3000/index.html#dialog2 

On other platforms (PC desktop, Browser) it works correctly. When I tried to create the dialog with completely different URL it worked also on Mac, but we would like to keep using the HashRouter.

**Comment by exextoc** [comment]
@schwarzwald We are unable to repro this at our end. Can you try the above sample code shared by us with your hash router URLs and share console logs and repro video?

**Comment by schwarzwald** [comment]
I have tried to run your code snippet with our hash router and this is the console output
```
2023-10-06 15:19:16.967    Dialog created
2023-10-06 15:19:16.968    Dialog close called
2023-10-06 15:19:16.999    new dialog created
2023-10-06 15:19:17.015    new dialog closed
```
Unfortunately I cannot share a video with you but I can describe what happens.

Actually nothing much happens at all. A dialog is opened showing the native Loading spinner and that's it. The dialog is not closed and the actual content is never loaded.

**Comment by exextoc** [comment]
So to summarise, this issue happens only when you use hash router and create dialogs back to back? Does adding delay between dialog creations help? Is it possible for you to share you Add-in so that we can try to repro the issue on our side? Do you see any error message in console in this case?

**Comment by schwarzwald** [comment]
> So to summarise, this issue happens only when you use hash router and create dialogs back to back?

Yes, exactly. I can reproduce the issue using your code snippet in combination with the hash router. But this is a private add-in so I can't share it with you.
I generated the project using the yeoman generator from the office add-in documentation.
The react app looks like this
```
const App: React.FC<AppProps> = (props: AppProps) => {
  return props.isOfficeInitialized ? (
    <HashRouter>
      <Switch>
        <Route path="/message">
          <MessageDialog />
        </Route>
        <Route path="/loading">
          <LoadingDialog />
        </Route>
      </Switch>
  </HashRouter>): (<></>);
}
```
Also no errors in console.

Adding a delay between dialog creation is our current workaround, so we are not entirely blocked by this.

**Comment by exextoc** [comment]
@schwarzwald Could you please share a sample add-in with us, so that we can repro the issue?

**Comment by schwarzwald** [comment]
I'm sorry I can't, the samples which I already provided is all I can do, I don't have the capacity to spend more time on this, we are currently using the workaround with timeout.
